### PR TITLE
Add support for more complex parameter domains

### LINKED ISSUE #63: Add a way for the user to provide more complex parameter domains
**Is your feature request related to a problem? Please describe.**
Right now, the parameter domain is limited to a hyper-rectangle. In many cases, this isn't sufficient as parameters might be subject to more complex conditions -- in particular, this is the case for the anisotropic heat model. 

**Describe the solution you'd like**
Instead of using the `param_limits` attribute of the model in `transformations.py` to determine whether the parameter is viable, we could add a small method to model called something like `is_within_param_limits(self, param)`. The default implementation in `model.py` simply checks if the parameter is within `param_limits`, but user models can provide their own implementation.

**Describe alternatives you've considered**
Theoretically, this could also be done via exception handling (at least if #62 gets implemented), maybe with a custom `ModelDomainError` exception, but I believe the approach described above is a lot clearer.

### PR DESCRIPTION
**Pull request by castellinilinguini**
# Description

Adds a function called `param_is_within_domain` to model that users can use in their models to specify more complex parameter domains. Fixes #63 .

## Type of change

Please tick at least one of the options. Delete the other options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

Please note/describe at least one tests that you ran to verify your changes.

- [x] Most unit tests
- [x] A small example in my jupyter notebook.

## Checklist For Contributor

- [x] My code follows the style guidelines of this project (I installed pre-commit)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I wrote my changes in the changelog in the `[unreleased]` section
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

## Checklist For Maintainers

- [ ] Continuous Integration (CI) is successfully running
- [ ] Do we want to release/tag a new version? [✔️ / ❌]
  - [ ] If yes, add a release to the changelog and set the new version in pyproject.toml. After the merge, tag the release!

### CODE REVIEW

**Comment by StudVingX** [comment]
Nice feature! lgtm

**Comment by TheVincentWagner** [comment]
Nice feature, lgtm!

File: eulerpi/core/model.py [hunk 1]
```diff
@@ -96,6 +96,21 @@ def forward_and_jacobian(
...
+        Overwrite this function if your model has a more complex parameter domain than a box.
+
+        Args:
+            param(np.ndarray): The parameter to check.
+
+        Returns:
+            bool: True if the parameter is within the limits.
+
+        """
+        return np.all(param >= self.param_limits[:, 0]) and np.all(
```
Comment (kaiserls) [inline eulerpi/core/model.py:110]:
```suggestion
               return True
```

The limits are already checked in `eulerpi/core/transformations.py`.
PS: I couldn't find a way to suggest changes spanning multiple rows.

File: eulerpi/core/model.py [hunk 2]
```diff
@@ -96,6 +96,21 @@ def forward_and_jacobian(
 
         return self.forward(param), self.jacobian(param)
 
+    def param_is_within_domain(self, param: np.ndarray) -> bool:
+        """Checks whether a parameter is within the parameter domain of the model.
+        Overwrite this function if your model has a more complex parameter domain than a box.
```
Comment (kaiserls) [inline eulerpi/core/model.py:101]: ... than a box. The param_limits are checked automatically.

You can link to the param_limits variable also.

**Comment by castellinilinguini** [comment]
The changes are already incorporated in the heat model branch, for which I will create a pull request shortly.

**Pull request closed without merging**